From 0f91f3583fc396259ae69d326d2e5694a5d64724 Mon Sep 17 00:00:00 2001
From: "yi.zhang1" <yi.zhang1@amlogic.com>
Date: Tue, 14 Nov 2023 10:55:27 +0800
Subject: [PATCH 08/14] transcoding: support video transcoding [1/1]

PD#SWPL-144371

Problem:
support video transcoding.

Solution:
add new bitbake file.

Verify:
local

Signed-off-by: yi.zhang1 <yi.zhang1@amlogic.com>
Change-Id: I2eff1fde0240aa81c07b9b056c2f85776d7af50b
---
 conf/machine/include/amlogic_enc_common.inc   |  9 ++++
 conf/machine/include/amlogic_externalsrc.inc  |  2 +
 conf/machine/mesong12b-w400-debian-5.15.conf  |  1 +
 .../machine/mesong12b-w400-weston-5.15.config |  2 +
 conf/machine/mesonsc2-ah212-debian-5.15.conf  |  1 +
 .../machine/mesonsc2-ah212-weston-5.15.config |  1 +
 conf/machine/mesonsm1-ac200-debian-5.15.conf  |  1 +
 .../machine/mesonsm1-ac200-weston-5.15.config |  1 +
 conf/machine/mesont7-an400-debian-5.15.conf   |  1 +
 conf/machine/mesont7-an400-debian.config      |  1 +
 conf/machine/mesont7-an400-weston-5.15.config |  1 +
 conf/machine/mesont7-an400-weston.config      |  1 +
 .../images/packagegroup-amlogic-baserootfs.bb |  2 +
 .../gst-aml-utils/gst-aml-videotranscoding.bb | 47 +++++++++++++++++++
 .../gst-aml-utils/videotranscoding-demo.bb    | 32 +++++++++++++
 15 files changed, 103 insertions(+)
 create mode 100644 conf/machine/include/amlogic_enc_common.inc
 create mode 100644 recipes-multimedia/gst-aml-utils/gst-aml-videotranscoding.bb
 create mode 100644 recipes-multimedia/gst-aml-utils/videotranscoding-demo.bb

diff --git a/conf/machine/include/amlogic_enc_common.inc b/conf/machine/include/amlogic_enc_common.inc
new file mode 100644
index 00000000..4cd4e667
--- /dev/null
+++ b/conf/machine/include/amlogic_enc_common.inc
@@ -0,0 +1,9 @@
+#@TYPE: Machine
+#@NAME: meson
+
+#@DESCRIPTION: Machine configuration for encoder
+
+CONNECTIVITY_CHECK_URIS = "https://www.yoctoproject.org/"
+
+DISTRO_FEATURES_append = " gst-aml-videotranscoding"
+DISTRO_FEATURES_append = " videotranscoding-demo "
\ No newline at end of file
diff --git a/conf/machine/include/amlogic_externalsrc.inc b/conf/machine/include/amlogic_externalsrc.inc
index d5847def..941ad0bd 100644
--- a/conf/machine/include/amlogic_externalsrc.inc
+++ b/conf/machine/include/amlogic_externalsrc.inc
@@ -153,6 +153,8 @@ EXTERNALSRC_pn-gst-plugin-nn= "${@'${MESON_ROOT_PATH}/aml-comp/multimedia/gst-am
 EXTERNALSRC_pn-detect-library= "${@'${MESON_ROOT_PATH}/aml-comp/multimedia/gst-aml-utils/detect-library' if (os.path.isdir('${MESON_ROOT_PATH}/aml-comp/multimedia/gst-aml-utils/detect-library')) else ''}"
 EXTERNALSRC_pn-gst-plugin-conv= "${@'${MESON_ROOT_PATH}/aml-comp/multimedia/gst-aml-utils/gst-plugin-conv' if (os.path.isdir('${MESON_ROOT_PATH}/aml-comp/multimedia/gst-aml-utils/gst-plugin-conv')) else ''}"
 EXTERNALSRC_pn-gst-aml-gfx2d= "${@'${MESON_ROOT_PATH}/aml-comp/multimedia/gst-aml-utils/gst-aml-gfx2d' if (os.path.isdir('${MESON_ROOT_PATH}/aml-comp/multimedia/gst-aml-utils/gst-aml-gfx2d')) else ''}"
+EXTERNALSRC_pn-gst-aml-videotranscoding= "${@'${MESON_ROOT_PATH}/aml-comp/multimedia/gst-aml-utils/gst-aml-videotranscoding' if (os.path.isdir('${MESON_ROOT_PATH}/aml-comp/multimedia/gst-aml-utils/gst-aml-videotranscoding')) else ''}"
+EXTERNALSRC_pn-videotranscoding-demo= "${@'${MESON_ROOT_PATH}/aml-comp/multimedia/gst-aml-utils/videotranscoding-demo' if (os.path.isdir('${MESON_ROOT_PATH}/aml-comp/multimedia/gst-aml-utils/videotranscoding-demo')) else ''}"
 EXTERNALSRC_pn-dma-allocator= "${@'${MESON_ROOT_PATH}/aml-comp/multimedia/gst-aml-utils/dma-allocator' if (os.path.isdir('${MESON_ROOT_PATH}/aml-comp/multimedia/gst-aml-utils/dma-allocator')) else ''}"
 EXTERNALSRC_pn-gst-aml-dma-allocator= "${@'${MESON_ROOT_PATH}/aml-comp/multimedia/gst-aml-utils/gst-aml-dma-allocator' if (os.path.isdir('${MESON_ROOT_PATH}/aml-comp/multimedia/gst-aml-utils/gst-aml-dma-allocator')) else ''}"
 EXTERNALSRC_pn-gst-jpeg-swdec= "${@'${MESON_ROOT_PATH}/aml-comp/multimedia/gst-aml-utils/gst-jpeg-swdec' if (os.path.isdir('${MESON_ROOT_PATH}/aml-comp/multimedia/gst-aml-utils/gst-jpeg-swdec')) else ''}"
diff --git a/conf/machine/mesong12b-w400-debian-5.15.conf b/conf/machine/mesong12b-w400-debian-5.15.conf
index decfb6c9..0d0edc15 100644
--- a/conf/machine/mesong12b-w400-debian-5.15.conf
+++ b/conf/machine/mesong12b-w400-debian-5.15.conf
@@ -8,6 +8,7 @@ include conf/machine/include/amlogic_weston.inc
 include conf/machine/include/amlogic_aiot_common.inc
 include conf/machine/include/amlogic_aiot_nndemo.inc
 include conf/machine/include/amlogic_debian_common.inc
+include conf/machine/include/amlogic_enc_common.inc
 
 PREFERRED_PROVIDER_virtual/mesa = "libgles-eabihf-gondul-wayland-drm"
 PREFERRED_PROVIDER_virtual/mesa-gl = "libgles-eabihf-gondul-wayland-drm"
diff --git a/conf/machine/mesong12b-w400-weston-5.15.config b/conf/machine/mesong12b-w400-weston-5.15.config
index 26c5af3c..361f1c1e 100644
--- a/conf/machine/mesong12b-w400-weston-5.15.config
+++ b/conf/machine/mesong12b-w400-weston-5.15.config
@@ -7,6 +7,8 @@ include conf/machine/include/mesong12b_k5.15_64b.inc
 include conf/machine/include/amlogic_weston.inc
 include conf/machine/include/amlogic_aiot_common.inc
 include conf/machine/include/amlogic_aiot_nndemo.inc
+include conf/machine/include/amlogic_enc_common.inc
+
 
 PREFERRED_PROVIDER_virtual/mesa = "libgles-eabihf-gondul-wayland-drm"
 PREFERRED_PROVIDER_virtual/mesa-gl = "libgles-eabihf-gondul-wayland-drm"
diff --git a/conf/machine/mesonsc2-ah212-debian-5.15.conf b/conf/machine/mesonsc2-ah212-debian-5.15.conf
index a12a3eec..92e99ab9 100644
--- a/conf/machine/mesonsc2-ah212-debian-5.15.conf
+++ b/conf/machine/mesonsc2-ah212-debian-5.15.conf
@@ -7,6 +7,7 @@ include conf/machine/include/mesonsc2_k5.15_64b.inc
 include conf/machine/include/amlogic_weston.inc
 include conf/machine/include/amlogic_aiot_common.inc
 include conf/machine/include/amlogic_debian_common.inc
+include conf/machine/include/amlogic_enc_common.inc
 
 PREFERRED_PROVIDER_virtual/mesa = "libgles-eabihf-dvalin-wayland-drm"
 PREFERRED_PROVIDER_virtual/mesa-gl = "libgles-eabihf-dvalin-wayland-drm"
diff --git a/conf/machine/mesonsc2-ah212-weston-5.15.config b/conf/machine/mesonsc2-ah212-weston-5.15.config
index 798ef7b9..d28cf175 100644
--- a/conf/machine/mesonsc2-ah212-weston-5.15.config
+++ b/conf/machine/mesonsc2-ah212-weston-5.15.config
@@ -7,6 +7,7 @@ include conf/machine/include/mesonsc2_k5.15_64b.inc
 include conf/machine/include/amlogic_weston.inc
 include conf/machine/include/amlogic_aiot_common.inc
 include conf/machine/include/amlogic_aiot_nndemo.inc
+include conf/machine/include/amlogic_enc_common.inc
 
 PREFERRED_PROVIDER_virtual/mesa = "libgles-eabihf-dvalin-wayland-drm"
 PREFERRED_PROVIDER_virtual/mesa-gl = "libgles-eabihf-dvalin-wayland-drm"
diff --git a/conf/machine/mesonsm1-ac200-debian-5.15.conf b/conf/machine/mesonsm1-ac200-debian-5.15.conf
index ff6c2cc7..03b13bd3 100644
--- a/conf/machine/mesonsm1-ac200-debian-5.15.conf
+++ b/conf/machine/mesonsm1-ac200-debian-5.15.conf
@@ -8,6 +8,7 @@ include conf/machine/include/amlogic_weston.inc
 include conf/machine/include/amlogic_aiot_common.inc
 include conf/machine/include/amlogic_aiot_nndemo.inc
 include conf/machine/include/amlogic_debian_common.inc
+include conf/machine/include/amlogic_enc_common.inc
 
 PREFERRED_PROVIDER_virtual/mesa = "libgles-eabihf-dvalin-wayland-drm"
 PREFERRED_PROVIDER_virtual/mesa-gl = "libgles-eabihf-dvalin-wayland-drm"
diff --git a/conf/machine/mesonsm1-ac200-weston-5.15.config b/conf/machine/mesonsm1-ac200-weston-5.15.config
index 74d7f433..cca5cb23 100644
--- a/conf/machine/mesonsm1-ac200-weston-5.15.config
+++ b/conf/machine/mesonsm1-ac200-weston-5.15.config
@@ -7,6 +7,7 @@ include conf/machine/include/mesonsm1_k5.15_64b.inc
 include conf/machine/include/amlogic_weston.inc
 include conf/machine/include/amlogic_aiot_common.inc
 include conf/machine/include/amlogic_aiot_nndemo.inc
+include conf/machine/include/amlogic_enc_common.inc
 
 PREFERRED_PROVIDER_virtual/mesa = "libgles-eabihf-dvalin-wayland-drm"
 PREFERRED_PROVIDER_virtual/mesa-gl = "libgles-eabihf-dvalin-wayland-drm"
diff --git a/conf/machine/mesont7-an400-debian-5.15.conf b/conf/machine/mesont7-an400-debian-5.15.conf
index 50444619..d7174562 100644
--- a/conf/machine/mesont7-an400-debian-5.15.conf
+++ b/conf/machine/mesont7-an400-debian-5.15.conf
@@ -8,6 +8,7 @@ include conf/machine/include/amlogic_weston.inc
 include conf/machine/include/amlogic_aiot_common.inc
 include conf/machine/include/amlogic_aiot_nndemo.inc
 include conf/machine/include/amlogic_debian_common.inc
+include conf/machine/include/amlogic_enc_common.inc
 
 PREFERRED_PROVIDER_virtual/mesa = "libgles-eabihf-gondul-wayland-drm"
 PREFERRED_PROVIDER_virtual/mesa-gl = "libgles-eabihf-gondul-wayland-drm"
diff --git a/conf/machine/mesont7-an400-debian.config b/conf/machine/mesont7-an400-debian.config
index c324ac68..9695dfb7 100644
--- a/conf/machine/mesont7-an400-debian.config
+++ b/conf/machine/mesont7-an400-debian.config
@@ -8,6 +8,7 @@ include conf/machine/include/amlogic_weston.inc
 include conf/machine/include/amlogic_aiot_common.inc
 include conf/machine/include/amlogic_aiot_nndemo.inc
 include conf/machine/include/amlogic_debian_common.inc
+include conf/machine/include/amlogic_enc_common.inc
 
 PREFERRED_PROVIDER_virtual/mesa = "libgles-eabihf-gondul-wayland-drm"
 PREFERRED_PROVIDER_virtual/mesa-gl = "libgles-eabihf-gondul-wayland-drm"
diff --git a/conf/machine/mesont7-an400-weston-5.15.config b/conf/machine/mesont7-an400-weston-5.15.config
index 269d47f4..68591a2d 100644
--- a/conf/machine/mesont7-an400-weston-5.15.config
+++ b/conf/machine/mesont7-an400-weston-5.15.config
@@ -7,6 +7,7 @@ include conf/machine/include/mesont7_k5.15_64b.inc
 include conf/machine/include/amlogic_weston.inc
 include conf/machine/include/amlogic_aiot_common.inc
 include conf/machine/include/amlogic_aiot_nndemo.inc
+include conf/machine/include/amlogic_enc_common.inc
 
 PREFERRED_PROVIDER_virtual/mesa = "libgles-eabihf-gondul-wayland-drm"
 PREFERRED_PROVIDER_virtual/mesa-gl = "libgles-eabihf-gondul-wayland-drm"
diff --git a/conf/machine/mesont7-an400-weston.config b/conf/machine/mesont7-an400-weston.config
index e4c950db..f1500d5c 100644
--- a/conf/machine/mesont7-an400-weston.config
+++ b/conf/machine/mesont7-an400-weston.config
@@ -7,6 +7,7 @@ include conf/machine/include/mesont7_k5.4_64b.inc
 include conf/machine/include/amlogic_weston.inc
 include conf/machine/include/amlogic_aiot_common.inc
 include conf/machine/include/amlogic_aiot_nndemo.inc
+include conf/machine/include/amlogic_enc_common.inc
 
 PREFERRED_PROVIDER_virtual/mesa = "libgles-eabihf-gondul-wayland-drm"
 PREFERRED_PROVIDER_virtual/mesa-gl = "libgles-eabihf-gondul-wayland-drm"
diff --git a/recipes-core/images/packagegroup-amlogic-baserootfs.bb b/recipes-core/images/packagegroup-amlogic-baserootfs.bb
index edb680eb..485be967 100644
--- a/recipes-core/images/packagegroup-amlogic-baserootfs.bb
+++ b/recipes-core/images/packagegroup-amlogic-baserootfs.bb
@@ -170,6 +170,8 @@ RDEPENDS_packagegroup-amlogic-baserootfs += " \
     ${@bb.utils.contains('DISTRO_FEATURES', 'detect-library', 'detect-library', '', d)} \
     ${@bb.utils.contains('DISTRO_FEATURES', 'gst-plugin-conv', 'gst-plugin-conv', '', d)} \
     ${@bb.utils.contains('DISTRO_FEATURES', 'gst-aml-gfx2d', 'gst-aml-gfx2d', '', d)} \
+    ${@bb.utils.contains('DISTRO_FEATURES', 'gst-aml-videotranscoding', 'gst-aml-videotranscoding', '', d)} \
+    ${@bb.utils.contains('DISTRO_FEATURES', 'videotranscoding-demo', 'videotranscoding-demo', '', d)} \
     ${@bb.utils.contains('DISTRO_FEATURES', 'dma-allocator', 'dma-allocator', '', d)} \
     ${@bb.utils.contains('DISTRO_FEATURES', 'gst-aml-dma-allocator', 'gst-aml-dma-allocator', '', d)} \
     ${@bb.utils.contains('DISTRO_FEATURES', 'gst-jpeg-swdec', 'gst-jpeg-swdec', '', d)} \
diff --git a/recipes-multimedia/gst-aml-utils/gst-aml-videotranscoding.bb b/recipes-multimedia/gst-aml-utils/gst-aml-videotranscoding.bb
new file mode 100644
index 00000000..d5e23d1f
--- /dev/null
+++ b/recipes-multimedia/gst-aml-utils/gst-aml-videotranscoding.bb
@@ -0,0 +1,47 @@
+SUMMARY = "amlogic gstreamer media player"
+
+LICENSE = "AMLOGIC"
+LIC_FILES_CHKSUM = "file://${COREBASE}/../meta-meson/license/AMLOGIC;md5=6c70138441c57c9e1edb9fde685bd3c8"
+
+DEPENDS += " gstreamer1.0 gstreamer1.0-plugins-base "
+
+RDEPENDS_${PN} = " "
+
+
+SRCREV ?="${AUTOREV}"
+PV = "${SRCPV}"
+
+#S = "${WORKDIR}/git/"
+inherit autotools pkgconfig
+
+do_configure[noexec] = "1"
+
+EXTRA_OEMAKE = "TARGET_DIR=${D} STAGING_DIR=${STAGING_DIR_TARGET} DESTDIR=${D} PKG_CONFIG=${STAGING_BINDIR_NATIVE}/pkg-config"
+
+do_compile(){
+    oe_runmake -C ${S} ${EXTRA_OEMAKE} all
+}
+
+do_install() {
+    install -d -m 0755 ${D}${libdir}
+    install -d -m 0755 ${D}${includedir}
+
+    install -D -m 0755 ${S}/*.so ${D}${libdir}/
+    install -D -m 0755 ${S}/inc/*.h ${D}${includedir}/
+
+    rm ${S}/*.so
+}
+
+
+
+#DEPENDS += "${@bb.utils.contains('EXTRA_OECONF', ' ', ' ', ' ', d)}"
+#RDEPENDS_${PN} += "${@bb.utils.contains('EXTRA_OECONF', ' ', ' ', ' ', d)}"
+
+#EXTRA_OEMAKE = "OUT_DIR=${B} CROSS=${TARGET_PREFIX} TARGET_DIR=${STAGING_DIR_TARGET} STAGING_DIR=${D} DESTDIR=${D}"
+
+
+FILES_${PN} = " ${libdir}/* "
+FILES_${PN}-dev = " ${includedir}/* \
+                    /usr/lib/pkgconfig/* \
+    "
+INSANE_SKIP_${PN} += "file-rdeps"
\ No newline at end of file
diff --git a/recipes-multimedia/gst-aml-utils/videotranscoding-demo.bb b/recipes-multimedia/gst-aml-utils/videotranscoding-demo.bb
new file mode 100644
index 00000000..55bea4ae
--- /dev/null
+++ b/recipes-multimedia/gst-aml-utils/videotranscoding-demo.bb
@@ -0,0 +1,32 @@
+DESCRIPTION = "Amlogic dma buffer allocator plugin"
+
+LICENSE = "AMLOGIC"
+LIC_FILES_CHKSUM = "file://${COREBASE}/../meta-meson/license/AMLOGIC;md5=6c70138441c57c9e1edb9fde685bd3c8"
+
+DEPENDS += "gstreamer1.0 gstreamer1.0-plugins-base"
+DEPENDS += "gst-aml-videotranscoding"
+RDEPENDS_${PN} += "gst-aml-videotranscoding"
+
+inherit autotools pkgconfig
+
+SRCREV ?="${AUTOREV}"
+
+do_configure[noexec] = "1"
+
+EXTRA_OEMAKE = "TARGET_DIR=${D} STAGING_DIR=${STAGING_DIR_TARGET} DESTDIR=${D} PKG_CONFIG=${STAGING_BINDIR_NATIVE}/pkg-config"
+
+do_compile(){
+    oe_runmake -C ${S} ${EXTRA_OEMAKE} all
+}
+
+
+do_install() {
+    install -d -m 0755 ${D}${bindir}
+    install -D -m 0755 ${S}/videotranscoding_demo ${D}${bindir}/
+    rm ${S}/videotranscoding_demo
+}
+
+FILES_${PN} = "${bindir}/*"
+FILES_${PN}-dev = "${includedir}/*"
+
+INSANE_SKIP_${PN} += "file-rdeps"
-- 
2.25.1

